code/datums/progress_bar.dm

```diff
@@ -0,0 +1,213 @@
+	holder.alpha = 0
+	holder.plane = HUD_PLANE
+	holder.appearance_flags = KEEP_TOGETHER | APPEARANCE_UI_IGNORE_ALPHA
+
+	backdrop = image(icon, target, "backdrop")
+
+	fill = image(icon, target, "fill")
+	fill.filters = filter(type = "alpha", icon = icon(icon, "mask"), x = -fill_width)
+
+	holder.overlays += backdrop
+	holder.overlays += fill
+
+/**
+  * Call periodically to update the progress bar's progress.
+  *
+  * Arguments:
+  * * progress - The new value, divided with the given goal on creation to calculate the fraction.
+  */
+/datum/progress_bar/proc/update(progress)
+	if(QDELETED(target))
+		qdel(src)
+		return
+
+	if(user?.client != client)
+		client?.images -= holder
+		if(!QDELETED(user) && user?.client)
+			client = user.client
+			client.images += holder
+
+	. = clamp(progress / goal, 0, 1)
+	last_progress = progress
+	if(fill)
+		holder.overlays -= fill
+		update_fill(.)
+		holder.overlays += fill
+
+/**
+  * Updates the fill. Do your visual changes here.
+  *
+  * Arguments:
+  * * fraction - The new (value / goal) fraction.
+  */
+/datum/progress_bar/proc/update_fill(fraction)
+	UNLINT(fill?.filters[1]?.x = -fill_width * (1 - fraction))
+
+/**
+  * Shifts the progress bar down.
+  */
+/datum/progress_bar/proc/shift()
+	animate(holder,
+		pixel_z = holder.pixel_z - height,
+		time = animation_time_shift,
+		easing = SINE_EASING | EASE_OUT,
+		flags = ANIMATION_PARALLEL
+	)
+
+/datum/progress_bar/default
+	name = "default"
+	icon = 'icons/effects/progress_bar/default.dmi'
+	fill_width = 22
+	height = 7
+
+/datum/progress_bar/default/update_fill(fraction)
+	. = ..()
+	if(!stopping)
+		fill.color = gradient(0, "#cc0033", 0.25, "#cc6633", 0.5, "#d1cc33", 0.75, "#00cc33", fraction)
+	else if(last_progress != goal)
+		animate(fill, time = 0.5 SECOND, loop = -1, flags = ANIMATION_PARALLEL, alpha = 0, color = "#cc0033")
+		animate(time = 0.5 SECOND, alpha = 255)
+	else
+		fill.color = COLOR_YELLOW
+
+/datum/progress_bar/default/slim
+	name = "default (slim)"
+	icon = 'icons/effects/progress_bar/default_slim.dmi'
+	height = 5
+
+/**
+  * An example progress bar that stays yellow then blinks red every now and then.
+  */
+/datum/progress_bar/warning
+	name = "warning"
+	icon = 'icons/effects/progress_bar/warning.dmi'
+	fill_width = 22
+	height = 7
+
+/datum/progress_bar/warning/update_fill(fraction)
+	. = ..()
+	if(fraction <= 0.5)
+		fill.color = COLOR_YELLOW
+	else
+		fill.color = gradient(0.5, COLOR_YELLOW, 0.5, COLOR_RED, "loop", fraction * 10)
+
+/datum/progress_bar/warning/slim
+	name = "warning (slim)"
+	icon = 'icons/effects/progress_bar/warning_slim.dmi'
+	height = 5
+
+/**
+  * Creates a new progress bar datum with the given theme.
+  *
+  * Arguments:
+  * * user - The mob seeing the progress bar.
+  * * goal - The goal number.
+  * * target - The atom to display the bar on.
+  * * theme - The progress bar's theme (type).
+  */
+/proc/create_progress_bar(mob/user, goal, atom/target, theme = /datum/progress_bar/default)
```

Comment:
What is the purpose of making this a global proc instead of just doing `new` like the older code?

Reply:
Creating by type is a bit hardcode. I want to left a way to pass different themes by preference choice.